### ISSUE TITLE
Question about library dependency unpack location

### ISSUE THREAD

**Issue by markjohndoyle** [maven-nar/nar-maven-plugin#239]
Hi all,

Just curious if there any technical reason(s) why we copy the dependencies to the target area and unpack them there, rather than just unpacking them in the .m2 directory and referencing that location?

If you have a large multi-module project you end up duplicating a lot of dependencies. Also, when you clean the project you lose all your dependencies until rebuild (or at least nar:nar-unpack-dependencies) which tends to upset IDES.

I suppose you've also got the inefficiency of copying and unpacking which wouldn't need to happen often if it was already carried out in the .m2.

**Comment by 00datman** [comment]
I've also wondered this.  It seems like we could unpack into the .m2 safely, right?

The biggest potential issue I'd foresee (with my limited foresight :) ) is that it might be confusing with -SNAPSHOT versions, but could we use the Nexus approach of expanding that into a dateTimeStamp?  Or leave it as -SNAPSHOT and know it's volatile?

**Comment by markjohndoyle** [comment]
It should unpack fine in the .m2, absolutely.

Snapshot are by definition volatile, yes - if you want a stable version you just depend upon a release. That's why the release plugin blocks you if you have snapshot dependencies.

I think the maven convention is to always take the latest snapshot, right? That must happen now, somehow, we just get the dependency, copy it, and unpack it. Maven must provide the latest snapshot. I think there are some metadata XML files in the repo that allow maven to handle this.

I don't know the details of the time stamp metadata system though :)

**Comment by GregDomjan** [comment]
Unpacking defaulted to local project space targetDirectory, this is mainly due to unpacking deleting existing unpacked folder before unpacking which makes it not thread or cross process safe.

If you only run maven once per each config, and don't run with -T threading options, you can configure with 
```
<configuration>
  ...
  <unpackDirectory>\some\shared\location</unpackDirectory>
```

https://github.com/maven-nar/nar-maven-plugin/blob/master/src/main/java/com/github/maven_nar/AbstractNarMojo.java#L122-133

Looks like validate is adding the defaults
  - Unclear why not using args within @Parameter to set default and possibly property to use

**Comment by ctrueden** [comment]
> this is mainly due to unpacking deleting existing unpacked folder before unpacking which makes it not thread or cross process safe.

Each Maven local repository cache is also not cross process safe; it is assumed that only one Maven instance is doing I/O in there at a time. So maybe using the local repo cache by default is a sensible idea?

**Comment by dproteus13** [comment]
I know that I've mostly been silent on here, and I apologize for that
(feel free to email me for explanation).  That said, I would love to see
this happen, but unfortunately am unable to make the change myself (again,
email if curious).  So, +1, and I'm sorry I can't do it myself!

**Comment by 00datman** [comment]
Forgive my naivete, I'm a C++ guy...  is this as easy as setting a @default for the unpackDirectory property?

**Comment by GregDomjan** [comment]
I think to default this value to anywhere but the local project doesn't have a good solution - besides relative path above the project location, how would we select any shared area.

Could allow alternate configuration from property by changing the Mojo 
@Parameter( property = "nar.unpackDirectory" )
@Parameter( property = "nar.testUnpackDirectory" )

**Comment by 00datman** [comment]
No, it is not :-P  I'm trying to implement this right now, stay tuned and say a prayer for my poor hurting C++ brain :-P

**Comment by 00datman** [comment]
Sorry, just saw your response...  what I tried just now was I copied a line from AbstractDependencyMojo that was along the lines of @property (default="${localRepository}"...

That broke the integration tests, though...

**Comment by 00datman** [comment]
@gregdomjan - I think that already works, though...  per your comment above, in <configuration> <unpackDirectory> lets me set it manually.   The question is, what value is there in defaulting to local build directory?  Why not default to local repository?

**Comment by GregDomjan** [comment]
@00datman  Working with the unpack directory is only thread safe currently if different directories are used for each project so they don't conflict in deleting/unpacking the same files.
There is some attempt to prevent multiple unpacking based on time stamp, it's not always successful.

If you have a reactor build and mvn compile -T2  then as any 2 projects are running one may delete and re unpack the same dependency - if unpack is same location this can break the other project already compiling/linking.

If  managing unpacking once within reactor and across threads was successful, then your suggestion of the localRepository (or something relative to it) would be safer as a general default.

**Comment by 00datman** [comment]
@gregdomjan Thank you for trying to help me understand, sorry if I'm a bit slow on the up-take!

I think what you're saying is that using any common location is a bad idea because native Maven behavior is not going to support that safely, given what NAR does? 

What if we changed things to alleviate that a bit, it seems like the issue is that we delete and then Re- unpack things. What if instead we skip deletion, and only unpack, which should (I think) not cause a conflict?

**Comment by markjohndoyle** [comment]
That as my question too, any idea why the unpacked dependency deleted at all? If it's a release version then then replacement that is unpacked it must be identical. If it's a snapshot the project should use the same latest snapshot from when the time the build started.

**Comment by GregDomjan** [comment]
@00datman @markjohndoyle   Just want to say I'd love to see some improvement here, not trying to say it's impossible, but rather it might take a more involved fix.

I think in most instances unpack if it's missing would be enough. 
Handling of RELEASE artifacts as if they are less volatile has not been attempted, seems reasonable on first thought at least.

If unpacking should happen, there are cases where not deleting first can lead to odd results 
 - file went missing, didn't notice because remained from last unpack
 - similarly, file deliberately removed still present, build failures not immediately apparent
 - need to ensure overwrite existing, can lead to partial results if files are locked.  Once files are changed folder timestamp is updated and so incomplete unpack appears to be sound.  Failure would need to include delete.

For RELEASE,  people modifying the unpacked folder, this may not only be intentional but sometimes say in an IDE header is opened and gets modified.   Readonly marking might help this.


If your running multi threaded
The check to see if each unpack needs to be update is being run in each thread for that modules dependencies. Being up to date is not shared, so it's possible for each thread to decide to unpack. 
Example
A depends on  CDEFG,  B depends on EFG
A and B running evaluate that CDEFG and EFG need updating
  A unpacks C, B unpack E, 
                        B unpack F, 
  A unpacks D, B unpacks G 
  B starts compiling, 
  A unpacks E, 
  A unpacks F, B fails to build because F is missing...
